### ISSUE TITLE
Discovered a possible fatal map / server killing bug last night

### ISSUE THREAD

**Issue by melindaendv** [EternalBlueFlame/Traincraft-5#171]
Apologizes I don't have a crash report mainly because I was in a frantic to try and save the server.. So I will try and explain what happened.

I had the track builder laying track and I have a freight car (US centerbeam 54 slot, the non wood products version) attached to the track layer to collect crap..
Track layer was filled with vanilla tracks coal blocks and ballast .

Everything was going fine until the track layer came to leaf level with a tree and it "Bounced" off and started coasting backwards. before I could stop it rolling backwards by removing it the server crashed.

I rebooted the server and tried to rejoin and nothing, it crashed. A friend of mine could rejoin but as soo as he got in chunk range of the track builder the server crashed.

How I saved my server was I downloaded the world and removed traincraft from my client and started it up in single player. All worked fine minus the missing traincraft notices and blocks in game..

I saved the world in single player and reuploaded it to my server and we spend hours manually replacing all the track and ballast on the map..

We are going to try and recreate this incident later on this evening on my test server to nail this problem down and get a crash report posted here. 

I have advised ALL my players to NOT use the track builder until further notice.

**Comment by NitroxydeX** [comment]
@melindaendv well. why no crash reports? Even when you save your server there is an extra folder for that?

**Comment by melindaendv** [comment]
let me see I think it may have been replaced

**Comment by melindaendv** [comment]
not sure how to add a spoiler or if thats a thing but it looks like it was that car attached that glitched... anyways here is the report I found.:

##Nitro-Note: added Spoiler
<details>
 <summary>

Crash Report</summary>


---- Minecraft Crash Report ----
// Why did you do that?

Time: 9/23/16 8:55 AM
Description: Ticking entity

java.lang.NullPointerException: Ticking entity
    at train.common.items.ItemTCRail.isTCTurnTrack(ItemTCRail.java:91)
    at train.common.api.EntityRollingStock.updateOnTrack(EntityRollingStock.java:1264)
    at train.common.api.EntityRollingStock.func_70071_h_(EntityRollingStock.java:861)
    at train.common.api.Freight.func_70071_h_(Freight.java:135)
    at net.minecraft.world.World.func_72866_a(World.java:2070)
    at net.minecraft.world.WorldServer.func_72866_a(WorldServer.java:648)
    at net.minecraft.world.World.func_72870_g(World.java:2034)
    at net.minecraft.world.World.func_72939_s(World.java:1887)
    at net.minecraft.world.WorldServer.func_72939_s(WorldServer.java:489)
    at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:636)
    at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:334)
    at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:547)
    at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:427)
    at net.minecraft.server.MinecraftServer$2.run(MinecraftServer.java:685)

## A detailed walkthrough of the error, its code path and all known details is as follows:

-- Head --
Stacktrace:
    at train.common.items.ItemTCRail.isTCTurnTrack(ItemTCRail.java:91)
    at train.common.api.EntityRollingStock.updateOnTrack(EntityRollingStock.java:1264)
    at train.common.api.EntityRollingStock.func_70071_h_(EntityRollingStock.java:861)
    at train.common.api.Freight.func_70071_h_(Freight.java:135)
    at net.minecraft.world.World.func_72866_a(World.java:2070)
    at net.minecraft.world.WorldServer.func_72866_a(WorldServer.java:648)
    at net.minecraft.world.World.func_72870_g(World.java:2034)

-- Entity being ticked --
Details:
    Entity Type: tc.Freight Center Beam Empty (train.common.entity.rollingStock.EntityFreightCenterbeam_Empty)
    Entity ID: 333566
    Entity Name: freightCenterBeam_Empty
    Entity's Exact location: 1770.28, 66.17, 954.50
    Entity's Block location: World: (1770,66,954), Chunk: (at 10,4,10 in 110,59; contains blocks 1760,0,944 to 1775,255,959), Region: (3,1; contains chunks 96,32 to 127,63, blocks 1536,0,512 to 2047,255,1023)
    Entity's Momentum: -0.03, 0.00, 0.00
Stacktrace:
    at net.minecraft.world.World.func_72939_s(World.java:1887)
    at net.minecraft.world.WorldServer.func_72939_s(WorldServer.java:489)

-- Affected level --
Details:
    Level name: 1474518051
    All players: 2 total; [GCEntityPlayerMP['matt8ln'/217566, l='1474518051', x=1168.13, y=67.15, z=954.50], GCEntityPlayerMP['Elfnet'/217924, l='1474518051', x=1776.98, y=64.99, z=952.94]]
    Chunk stats: ServerChunkCache: 710 Drop: 0
    Level seed: -3069899127109662574
    Level generator: ID 05 - BIOMESOP, ver 0. Features enabled: true
    Level generator options: 
    Level spawn location: World: (904,64,923), Chunk: (at 8,4,11 in 56,57; contains blocks 896,0,912 to 911,255,927), Region: (1,1; contains chunks 32,32 to 63,63, blocks 512,0,512 to 1023,255,1023)
    Level time: 823049 game time, 1638240 day time
    Level dimension: 0
    Level storage version: 0x04ABD - Anvil
    Level weather: Rain time: 25617 (now: false), thunder time: 123102 (now: false)
    Level game mode: Game mode: survival (ID 0). Hardcore: false. Cheats: false
Stacktrace:
    at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:636)
    at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:334)
    at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:547)
    at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:427)
    at net.minecraft.server.MinecraftServer$2.run(MinecraftServer.java:685)

-- System Details --
Details:
    Minecraft Version: 1.7.10
    Operating System: Linux (amd64) version 3.10.0-327.28.2.el7.x86_64
    Java Version: 1.8.0_101, Oracle Corporation
    Java VM Version: OpenJDK 64-Bit Server VM (mixed mode), Oracle Corporation
    Memory: 448146080 bytes (427 MB) / 1267867648 bytes (1209 MB) up to 3113877504 bytes (2969 MB)
    JVM Flags: 6 total; -Xms32M -Xmx3072M -XX:+UseParNewGC -XX:ParallelGCThreads=4 -XX:+CMSIncrementalPacing -XX:+CMSClassUnloadingEnabled
    AABB Pool Size: 0 (0 bytes; 0 MB) allocated, 0 (0 bytes; 0 MB) used
    IntCache: cache: 0, tcache: 0, allocated: 12, tallocated: 97
    FML: MCP v9.05 FML v7.10.99.99 Minecraft Forge 10.13.4.1614 80 mods loaded, 71 mods active
    States: 'U' = Unloaded 'L' = Loaded 'C' = Constructed 'H' = Pre-initialized 'I' = Initialized 'J' = Post-initialized 'A' = Available 'D' = Disabled 'E' = Errored
    UCHIJAAAA   mcp{9.05} [Minecraft Coder Pack](minecraft.jar) 
    UCHIJAAAA   FML{7.10.99.99} [Forge Mod Loader](minecraft_server.jar) 
    UCHIJAAAA   Forge{10.13.4.1614} [Minecraft Forge](minecraft_server.jar) 
    UCHIJAAAA   CodeChickenCore{1.0.4.35} [CodeChicken Core](minecraft.jar) 
    UCHIJAAAA   Micdoodlecore{} [Micdoodle8 Core](minecraft.jar) 
    UCHIJAAAA   NotEnoughItems{1.0.4.90} [Not Enough Items](not-enough-items-1.7.10-1.0.4.90.jar) 
    UCHIJAAAA   {000} [CoFH ASM](minecraft.jar) 
    UCHIJAAAA   {0} [DragonAPI ASM Data Initialization](minecraft.jar) 
    UCHIJAAAA   CoFHCore{1.7.10R3.1.3} [CoFH Core](CoFHCore-[1.7.10]3.1.3-327.jar) 
    UCHIJAAAA   BuildCraft|Core{7.1.18} [BuildCraft](buildcraft-7.1.18.jar) 
    UCHIJAAAA   BuildCraft|Transport{7.1.18} [BC Transport](buildcraft-7.1.18.jar) 
    UCHIJAAAA   BuildCraft|Silicon{7.1.18} [BC Silicon](buildcraft-7.1.18.jar) 
    UCHIJAAAA   BuildCraft|Energy{7.1.18} [BC Energy](buildcraft-7.1.18.jar) 
    UCHIJAAAA   BuildCraft|Factory{7.1.18} [BC Factory](buildcraft-7.1.18.jar) 
    UCHIJAAAA   BuildCraft|Builders{7.1.18} [BC Builders](buildcraft-7.1.18.jar) 
    UCHIJAAAA   ThermalFoundation{1.7.10R1.2.5} [Thermal Foundation](ThermalFoundation-[1.7.10]1.2.5-115.jar) 
    UCHIJAAAA   ThermalExpansion{1.7.10R4.1.4} [Thermal Expansion](ThermalExpansion-[1.7.10]4.1.4-247.jar) 
    UCHIJAAAA   Additional-Buildcraft-Objects{MC1.7.10-BC7.1release4.0.6} [Additional Buildcraft Objects](ABO-MC1.7.10-BC7.1release4.0.6.jar) 
    UCHIJAAAA   BigReactors{0.4.2A2} [Big Reactors](big-reactors-0.4.2A2.jar) 
    UCHIJAAAA   BiomesOPlenty{2.1.0} [Biomes O' Plenty](BiomesOPlenty-1.7.10-2.1.0.1889-universal.jar) 
    UCHIJAAAA   BuildCraft|Robotics{7.1.18} [BC Robotics](buildcraft-7.1.18.jar) 
    UCHIJAAAA   CarpentersBlocks{3.3.7} [Carpenter's Blocks](Carpenter's Blocks v3.3.7 - MC 1.7.10.jar) 
    UCHIJAAAA   IC2{2.2.657-experimental} [IndustrialCraft 2](Industrial-Craft-2-Mod-Experimental-1.7.10.jar) 
    UCHIJAAAA   Forestry{4.0.7.35} [Forestry for Minecraft](forestry_1.7.10-4.0.7.35.jar) 
    UCHIJAAAA   ForgeMultipart{1.2.0.345} [Forge Multipart](ForgeMultipart-1.7.10-1.2.0.345-universal.jar) 
    UCHIJAAAA   MrTJPCoreMod{1.1.0.33} [MrTJPCore](MrTJPCore-1.7.10-1.1.0.33-universal.jar) 
    UCHIJAAAA   ProjRed|Core{4.7.0pre12.95} [ProjectRed Core](ProjectRed-1.7.10-4.7.0pre12.95-Base.jar) 
    UCHIJAAAA   DragonAPI{1.0} [DragonAPI](DragonAPI 1.7.10 V15a.jar) 
    UCHIJAAAA   Eln{51} [Electrical Age](ElectricalAge_BETA-1.11_r51.jar) 
    UCHIJAAAA   MineFactoryReloaded{1.7.10R2.8.1} [MineFactory Reloaded](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UCHIJAAAA   EnderIO{1.7.10-2.2.8.381} [Ender IO](EnderIO-1.7.10-2.2.8.381.jar) 
    UCHIJAAAA   factorization.notify{1.0} [Factorization Notification System](Factorization-1.7.10-0.8.101.jar) 
    UCHIJAAAA   factorization.dimensionalSlice{0.8.101} [Factorization Dimensional Slices](Factorization-1.7.10-0.8.101.jar) 
    UCHIJAAAA   factorization{0.8.101} [Factorization](Factorization-1.7.10-0.8.101.jar) 
    UCHIJAAAA   factorization.misc{0.8.101} [Factorization Miscellaneous Nonsense](Factorization-1.7.10-0.8.101.jar) 
    UCHIJAAAA   fz.scrap{0.8.101} [Scrap](Factorization-1.7.10-0.8.101.jar) 
    UCHIJAAAA   factorization.truth{0.8.101} [Truth](Factorization-1.7.10-0.8.101.jar) 
    UCHIJAAAA   GalacticraftCore{3.0.12} [Galacticraft Core](GalacticraftCore-1.7-3.0.12.463.jar) 
    UCHIJAAAA   GalacticraftMars{3.0.12} [Galacticraft Planets](Galacticraft-Planets-1.7-3.0.12.463.jar) 
    UCHIJAAAA   IC2NuclearControl{2.1.2a} [Nuclear Control 2](ic2-nuclear-control-2.1.2a.jar) 
    UCHIJAAAA   Railcraft{9.12.2.0} [Railcraft](Railcraft_1.7.10-9.12.2.0.jar) 
    UCHIJAAAA   ImmersiveEngineering{0.7.7} [Immersive Engineering](ImmersiveEngineering-0.7.7.jar) 
    UCHIJAAAA   IronChest{6.0.62.742} [Iron Chest](ironchest-1.7.10-6.0.62.742-universal.jar) 
    UCHIJAAAA   MineFactoryReloaded|CompatBuildCraft{1.7.10R2.8.1} [MFR Compat: BuildCraft](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UCHIJAAAA   MineFactoryReloaded|CompatForestry{1.7.10R2.8.1} [MFR Compat: Forestry](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UCHIJAAAA   MineFactoryReloaded|CompatForgeMicroblock{1.7.10R2.8.1} [MFR Compat: ForgeMicroblock](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UCHIJAAAA   MineFactoryReloaded|CompatIC2{1.7.10R2.8.1} [MFR Compat: IC2](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UCHIJAAAA   ProjRed|Exploration{4.7.0pre12.95} [ProjectRed Exploration](ProjectRed-1.7.10-4.7.0pre12.95-World.jar) 
    UCHIJAAAA   MineFactoryReloaded|CompatProjRed{1.7.10R2.8.1} [MFR Compat ProjectRed](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UCHIJAAAA   MineFactoryReloaded|CompatRailcraft{1.7.10R2.8.1} [MFR Compat: Railcraft](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UCHIJAAAA   MineFactoryReloaded|CompatThermalExpansion{1.7.10R2.8.1} [MFR Compat: Thermal Expansion](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UCHIJAAAA   MineFactoryReloaded|CompatVanilla{1.7.10R2.8.1} [MFR Compat: Vanilla](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UCHIJAAAA   Monoblocks{1.10.33} [�1M�2o�3n�4o�5b�6l�7o�8c�9k�as](Monoblocks-1.10.33.jar) 
    UCHIJAAAA   MonoblocksMultipart{1.10.33} [MonoblocksMultipart](Monoblocks-1.10.33.jar) 
    UCHIJAAAA   movillages{1.2.1} [Mo' Villages](MoVillages-1.3.1.jar) 
    UCHIJAAAA   cfm{3.4.7} [�9MrCrayfish's Furniture Mod](MrCrayfishFurnitureModv3.4.7%281.7.10%29.jar) 
    UCHIJAAAA   ProjRed|Transmission{4.7.0pre12.95} [ProjectRed Transmission](ProjectRed-1.7.10-4.7.0pre12.95-Integration.jar) 
    UCHIJAAAA   ProjRed|Transportation{4.7.0pre12.95} [ProjectRed Transportation](ProjectRed-1.7.10-4.7.0pre12.95-Mechanical.jar) 
    UCHIJAAAA   ProjRed|Compatibility{4.7.0pre12.95} [ProjectRed Compatibility](ProjectRed-1.7.10-4.7.0pre12.95-Compat.jar) 
    UCHIJAAAA   ProjRed|Integration{4.7.0pre12.95} [ProjectRed Integration](ProjectRed-1.7.10-4.7.0pre12.95-Integration.jar) 
    UCHIJAAAA   ProjRed|Fabrication{4.7.0pre12.95} [ProjectRed Fabrication](ProjectRed-1.7.10-4.7.0pre12.95-Fabrication.jar) 
    UCHIJAAAA   ProjRed|Illumination{4.7.0pre12.95} [ProjectRed Illumination](ProjectRed-1.7.10-4.7.0pre12.95-Lighting.jar) 
    UCHIJAAAA   ProjRed|Expansion{4.7.0pre12.95} [ProjectRed Expansion](ProjectRed-1.7.10-4.7.0pre12.95-Mechanical.jar) 
    UCHIJAAAA   ThermalDynamics{1.7.10R1.2.0} [Thermal Dynamics](ThermalDynamics-[1.7.10]1.2.0-171 %281%29.jar) 
    UCHIJAAAA   tc{4.3.0_004} [Traincraft](Traincraft-4.3.0_004.jar) 
    UCHIJAAAA   worldedit{6.1.1} [WorldEdit](worldedit-forge-mc1.7.10-6.1.1-dist.jar) 
    UCHIJAAAA   McMultipart{1.2.0.345} [Minecraft Multipart Plugin](ForgeMultipart-1.7.10-1.2.0.345-universal.jar) 
    UCHIJAAAA   ForgeRelocation{0.0.1.4} [ForgeRelocation](ForgeRelocation-1.7.10-0.0.1.4-universal.jar) 
    UCHIJAAAA   MCFrames{1.0} [MCFrames](ForgeRelocation-1.7.10-0.0.1.4-universal.jar) 
    UCHIJAAAA   RelocationFMP{0.0.1.2} [RelocationFMP](ForgeRelocationFMP-1.7.10-0.0.1.2-universal.jar) 
    UCHIJAAAA   ForgeMicroblock{1.2.0.345} [Forge Microblocks](ForgeMultipart-1.7.10-1.2.0.345-universal.jar) 
    UD  MineFactoryReloaded|CompatAppliedEnergistics{1.7.10R2.8.1} [MFR Compat: Applied Energistics](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UD  MineFactoryReloaded|CompatAtum{1.7.10R2.8.1} [MFR Compat: Atum](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UD  MineFactoryReloaded|CompatBackTools{1.7.10R2.8.1} [MFR Compat: BackTools](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UD  MineFactoryReloaded|CompatChococraft{1.7.10R2.8.1} [MFR Compat: Chococraft](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UD  MineFactoryReloaded|CompatExtraBiomes{1.7.10R2.8.1} [MFR Compat: ExtraBiomes](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UD  MineFactoryReloaded|CompatSufficientBiomes{1.7.10R2.8.1} [MFR Compat: Sufficient Biomes](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UD  MineFactoryReloaded|CompatThaumcraft{1.7.10R2.8.1} [MFR Compat: Thaumcraft](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UD  MineFactoryReloaded|CompatTConstruct{1.7.10R2.8.1} [MFR Compat: Tinkers' Construct](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    UD  MineFactoryReloaded|CompatTwilightForest{1.7.10R2.8.1} [MFR Compat: TwilightForest](MineFactoryReloaded-[1.7.10]2.8.1-174.jar) 
    CoFHCore: -[1.7.10]3.1.3-327
    ThermalFoundation: -[1.7.10]1.2.5-115
    ThermalExpansion: -[1.7.10]4.1.4-247
    MineFactoryReloaded: -[1.7.10]2.8.1-174
    ThermalDynamics: -[1.7.10]1.2.0-171
    Profiler Position: N/A (disabled)
    Vec3 Pool Size: 0 (0 bytes; 0 MB) allocated, 0 (0 bytes; 0 MB) used
    Player Count: 2 / 70; [GCEntityPlayerMP['matt8ln'/217566, l='1474518051', x=1168.13, y=67.15, z=954.50], GCEntityPlayerMP['Elfnet'/217924, l='1474518051', x=1776.98, y=64.99, z=952.94]]
    Is Modded: Definitely; Server brand changed to 'fml,forge'
    Type: Dedicated Server (map_server.txt) 

</details>

**Comment by MggMuggins** [comment]
That's a lot of extra mods...

**Comment by melindaendv** [comment]
yeah I know, They been working fine though, no problems except that one incident last night, it hasn't happened again..

**Comment by NitroxydeX** [comment]
well. The crash is from TC. It happens to the center beam freight cart. it crashed while detecting if this was a curve where it was going. thats... well. Do you know over which track the cart went before the crash?

**Comment by NitroxydeX** [comment]
basically you can remove erroring entities with a forge function. you just need to enable it in the forge config.

**Comment by melindaendv** [comment]
Yeah I was looking at that too. The track was being laid by the track builder and the center beam car was attached to it to pick up the mined blocks so all the track was straight..

It may have been while it was coasting backwards, I was replacing the vanilla track with TC track but I don't think I missed a spot and the car derailed and got confused.. It was late when it happened though and I was far behind the track builder also. Maybe it didn't lay a section for some reason and it hit that while rolling backwards?

Anyways I will investigate on my test server in a little while and see if thats the case

**Comment by NitroxydeX** [comment]
well... you cant place TC Tracks with the track builder.

**Comment by melindaendv** [comment]
Oh I know, I had it placing vanilla track xD I was going behind it manually than placing the TC track, basically I was using the track builder to bore tunnels and basic bridge deck. Alot faster than mining it by hand.

**Comment by NitroxydeX** [comment]
Well I think it tried to remove a TC Track which is why we say don't use it on TC Tracks. ^^

**Comment by melindaendv** [comment]
The traincraft track was quite a few blocks behind the builder train a good 100 blocks to be sure I was not interfering with its building operation.. What we discovered after replicating the situation was the builder glitched on a tree at leaf level, and for a second went sideways and laid a track 90 degrees to the rest of the track, then the builder lined back up and continued on in the in the planned direction, that one piece of sideways placed track when the centerbeam car went over it caused the whole train to start coasting backwards and then game crash..

My solution for this was attach a vanilla chest cart or no cart at all instead of a TC car to the builder and try to clear trees at leaf level ahead of the builder. This has worked so far an no more crashes.

**Comment by NitroxydeX** [comment]
the tree glitch should be removed in the next release.

**Comment by EternalBlueFlame** [comment]
Forgot to add a comment here before.
I ran into this same crash once, the issue was actually a TC track's variable didn't get set properly during re-spawn, so when an actual rollingstock tried to roll over it, it took a crap because it didn't know what it just rolled over.

The builder didn't cause the crash, it doesn't extend freight at any point, and whatever made the crash, extended freight, but the Center Beam Freight car, which was listed as the entity causing the crash, does.

This commit I made a while ago may fix the track being null, which is what caused the crash.
There was another commit I made which was also related and may help ensure its fixed, but I cant remember which one for the life of me.
[https://github.com/EternalBlueFlame/Traincraft-5/commit/baf50ddd6dba7a909987d0519540a9e30fde729b](https://github.com/EternalBlueFlame/Traincraft-5/commit/baf50ddd6dba7a909987d0519540a9e30fde729b)
